src/util/invariant.h

```diff
@@ -51,14 +54,64 @@ Author: Martin Brain, martin.brain@diffblue.com
 ** CPROVER_INVARIANT_* macros.
 */
 
+/// A logic error, augmented with a distinguished field to hold a backtrace.
 class invariant_failedt: public std::logic_error
 {
+private:
+  /// Stores the backtrace set by `invariant_violated`.
+  std::string backtrace;
+
+  /// Exposes backtrace directly to the violated invariant dispatch method:
+  template<class T> friend
+  typename std::enable_if<std::is_base_of<invariant_failedt, T>::value>::type
+  invariant_violated(
+    const std::string &file,
+    const std::string &function,
+    const int line,
+    const T reason);
+
+  /// Temporary storage for `what`'s return value
+  mutable std::string what_str;
+
 public:
-  explicit invariant_failedt(const std::string& what) : logic_error(what) {}
-  explicit invariant_failedt(const char* what) : logic_error(what) {}
+  invariant_failedt(
+    const std::string &what):
+    logic_error(what)
+  {
+  }
+
+  /// Return a human-readable description of this error.
+  /// For this base class that's just the description string
+  /// supplied to the constructor; subclasses may override this to
+  /// provide a description in a different way, e.g. from a more
+  /// structured error description. `std::logic_error::what` is
+  /// overridden to use this and the backtrace to describe the
+  /// error.
+  virtual std::string pretty_print() const
```

Comment:
Exceptions already have a virtual const function which returns a string, so I'm not convinced there's a need for another. I would restructure this code like so:

```c++
class invariant_failedt : public std::logic_error {
  // We don't need a mutable string holding the return of `what`, so the
  // object is smaller *and* more const-safe
  std::string backtrace;
public:
  // This is not virtual! It will call `what` and add a stack trace
  std::string pretty_print() const {
    try {
      std::ostringstream ss;
      ss << what() << '\n' + backtrace;
      return ss.str();
    } catch (...) {
      return "Exception describing invariant failure";
    }
  }

  // We don't even need to override `what` here, but derived classes are free
  // to do so if they need more structured info.
};
```

A possible downfall of this approach is that clients need to know to call `pretty_print` when they catch the exception,
rather than `what`, but I believe this is a reasonable trade-off, as the invariant_failedt no longer needs to store the what_str member, making it smaller and more const-safe.